<a href="https://colab.research.google.com/github/CodingDuckmx/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Jesus_CaballeroMedrano_LS_DS_123_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 123

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).



In [0]:
# TODO - code!

def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):
    new_prob_drunk_prior = (true_positive_rate * prob_drunk_prior) / ((true_positive_rate * prob_drunk_prior) + (false_positive_rate*(1-prob_drunk_prior)))
    return new_prob_drunk_prior





In [9]:
prob_drunk_given_positive(0.001,0.08,1)

0.012357884330202669

The probability that a person is drunk after one positive breathlyzer test is 1.2357%.

In [41]:
prob_drunk_given_positive(0.012357884330202669,0.08,1)

0.13525210993291495

The probability that a person is drunk after two positive breathlyzer test is 13.5252%

In [40]:
### How many test are needed to have a probability greater than 95% that a person is drunk beyond the legal limit?

prob_drunk_prior =  1/1000
i=0

while prob_drunk_prior < 0.95:
  prob_drunk_prior = prob_drunk_given_positive(prob_drunk_prior,0.08,1)
  i= i+1
  
else:  
  print(i)
  print(prob_drunk_prior)



4
0.9606895076105054


We need 4 test to reach that confidence.

###Explore scipy.stats.bayes_mvs

In [2]:
#Import the database

from google.colab import files
uploaded = files.upload()


Saving bank-full.csv to bank-full.csv


In [0]:
import pandas as pd
import numpy as np
from scipy import stats

In [46]:
df = pd.read_csv('bank-full.csv', sep=';')

df.head()

age           job  marital  education  ... pdays  previous poutcome   y
0   58    management  married   tertiary  ...    -1         0  unknown  no
1   44    technician   single  secondary  ...    -1         0  unknown  no
2   33  entrepreneur  married  secondary  ...    -1         0  unknown  no
3   47   blue-collar  married    unknown  ...    -1         0  unknown  no
4   33       unknown   single    unknown  ...    -1         0  unknown  no

[5 rows x 17 columns]

We want to know a confidence interval for the mean of the call duration.

In [0]:
def confidence_interval(data,confidence=0.95):
  n=len(data)
  mean = sum(data) / n
  data = np.array(data)
  sterr = stats.sem(data)
  interval = sterr * stats.t.ppf((1+ confidence) / 2.0, n-1)
  return (mean / 2.0, mean - interval, mean + interval)

In [59]:
confidence_interval(df['duration'])

(129.08153989073455, 255.78918317261343, 260.53697639032475)

In [42]:
### Bayesian approach

bayesian_confidence_interval, _, _ =stats.bayes_mvs(df['duration'], alpha=0.95)
bayesian_confidence_interval

Mean(statistic=258.1630797814691, minmax=(255.78927297965168, 260.5368865832865))

In [57]:
##### DRAFT ONLY ####


# yloan = df[df['loan']== 'yes'][['age']]

# nloan = df[df['loan']== 'no'][['age']]

# print(yloan.head())

# print(nloan.head())

np.array(df['duration'])

len(df['duration'])
sum(df['duration']) / n
  # data = np.array(data)
stats.sem(df['duration'])

1.2111615690458226

##Difference between Bayesian and frequentist statistics. 

Frequentist Statistics are based of probabilities build of frecuency of random events.

Bayesian Statistics are statistics updated everytime new information comes along.

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!